hazelcast/src/main/java/com/hazelcast/internal/memory/impl/PlatformAwareMemoryAccessor.java

```diff
@@ -0,0 +1,430 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.memory.impl;
+
+import com.hazelcast.internal.memory.MemoryAccessor;
+
+import java.lang.reflect.Field;
+
+/**
+ * <p>
+ * Platform aware {@link MemoryAccessor} that checks underlying platform and behaves regarding to its architecture
+ * as aligned or standard.
+ * </p>
+ *
+ * <p>
+ * If the underlying platform supports unaligned memory accesses,
+ * it behaves as standard {@link MemoryAccessor}  because no need to additional checks.
+ * Otherwise, it behaves as alignment-aware {@link MemoryAccessor} and
+ * checks and handles unaligned memory accesses if possible.
+ * </p>
+ *
+ * <p>
+ * But on the other hand, even though underlying platform supports unaligned memory accesses,
+ * unaligned memory accesses might cause performance overhead.
+ * But I am not sure at the moment it is really a problem on modern CPUs.
+ * So we have standard {@link MemoryAccessor} for these platforms such as `x86`, `x64, etc ...
+ * </p>
+ */
+public class PlatformAwareMemoryAccessor implements MemoryAccessor {
```

Comment:
Can this be moved into `MemoryAccessorProvider`? (since we have basically to `MemoryAccessor`)